adobe/copy_on_write.hpp

```diff
@@ -55,93 +36,98 @@ copy when the value is modified and there is more than one reference to the
+
+        T _value;
+    };
+
+    void destroy() {
+        if (_self && (--_self->_count == 0)) delete _self;
+    }
+
+    model* _self;
+
+    template <class U>
+    using disable_copy = std::enable_if_t<!std::is_same<std::decay_t<U>, copy_on_write>::value>*;
+
+    template <typename U>
+    using disable_copy_assign =
+        std::enable_if_t<!std::is_same<std::decay_t<U>, copy_on_write>::value, copy_on_write&>;
 #endif
 
+public:
+    /// \brief The type of value stored.
+    /* [[deprecated]] */ using value_type = T;
+
+    /// \brief The type of value stored.
+    using element_type = T;
+
     /*!
         The first call to the default constructor will construct a default
-        instance of value_type which will be used for subsequent calls to the
+        instance of element_type which will be used for subsequent calls to the
         default constructor. The default instance will be released at exit.
     */
     copy_on_write() {
-        std::call_once(flag_s, init_default);
-        object_m = default_s;
-        object_m->header_m.get().count_m.increment();
-    }
-
-    /*!
-        Initializes the instance with a custom allocator
-    */
-    explicit copy_on_write(const allocator_type& a) : object_m(0) {
-        implementation_allocator_type other_allocator(a);
-
-        object_m = allocate(other_allocator);
+        static model default_s;
+        _self = &default_s;
+        ++_self->_count;
     }
 
     /*!
     \brief Constructs a new copy_on_write object with a value \c x.
 
     \param x A default value to assign to this object
     */
-    copy_on_write(T x) : object_m(allocate_move(0, std::move(x))) {}
+    template <class U>
+    copy_on_write(U&& x, disable_copy<U> = nullptr) : _self(new model(std::forward<U>(x))) {}
 
     /*!
     Copy construction is a non-throwing operation and simply increments the
     reference count on the stored object.
     */
-    copy_on_write(const copy_on_write& x) : object_m(x.object_m) {
-        if (object_m)
-            object_m->header_m.get().count_m.increment();
-    }
+    copy_on_write(const copy_on_write& x) noexcept : _self(x._self) { ++_self->_count; }
+    copy_on_write(copy_on_write&& x) noexcept : _self(x._self) { x._self = nullptr; }
 
-    copy_on_write(copy_on_write&& x) : object_m(x.object_m) { x.object_m = 0; }
-
-    ~copy_on_write() { release(object_m); }
+    ~copy_on_write() { destroy(); }
 
     /*!
         As with copy construction, assignment is a non-throwing operation which
         releases the old value and increments the reference count of the item
         being assigned to.
     */
-    copy_on_write& operator=(copy_on_write x) {
-        swap(*this, x);
+    auto operator=(const copy_on_write& x) noexcept -> copy_on_write& {
+        return *this = copy_on_write(x);
+    }
+    auto operator=(copy_on_write&& x) noexcept -> copy_on_write& {
+        assert(!identity(x) && "FATAL (sparent) : invalid self-move.");
```

Comment:
[ I hate that GitHub will mark comments as outdated, that I haven't seen yet - I want it to have a "close comment" feature. Caught this comment only because it was in my email. ] Your comment jogged my memory on something Howard had written - that the aliasing problem goes beyond self assignment if you have anything that could be self referential. I went back through that discussion and put in a better fix. Howard's argument was that the self check isn't sufficient for any structure that is self referential - though it doesn't apply here (because value semantics) his fix does and is a great improvement in the code. This will still assert if rhs is moved from, but allows self move.